gms/gossiper.hh

```diff
@@ -87,6 +87,8 @@ class feature_service;
 struct bind_messaging_port_tag {};
 using bind_messaging_port = bool_class<bind_messaging_port_tag>;
 
+using advertise_myself_t = bool_class<class advertise_myself_tag>;
```

Comment:
nit: don't need the `_t` suffix.
seems to go against the conventional use of `bool_class` elsewhere.
Even, `bind_messaging_port` right above.

Reply:
> > > > This solution introduces a new mechanism in the gossiper which is supposed to be the same in its effects as the long shadow round, but rather than suppressing message sending it relies on the fact that generation zero will be ignored by existing nodes.
> > > > `gossiper::do_apply_state_locally()` will indeed ignore generations older than the one which is already learned. But what if the other node forgot it and haven't learnt it yet? Looking at the code, I think it will accept the zero generation in the `!es` path.
> > > 
> > > 
> > > How would a node ignore existing node in normal status in endpoint_state_map? Even if the node node restarts, it will inject an entry in endpoint_state_map.
> > 
> > 
> > When the node restarts, it will indeed inject an entry into the endpoint_state_map (storage_service does this) with generation 0, which is equal to the "shadow" generation used by this patch, so "shadow" gossip will be accepted, not ignored as we want.
> > Secondly, it's not the gossiper component who injects those entries, but scylla's storage service. It's better if gossiper's `advertise_myself::no` is not coupled with external things like that because they can change independently of the gossiper. Suppressing messages within the gossiper doesn't rely on external things like this, so it's better.
> 
> Good that you agree that the !es path is not possible as you claimed earlier. The node will run the remote_generation == local_generation path. However, we do not even need storage service to inject the zero generation to make it work. No node in a working state will have generation zero, such node should never be marked as alive by other nodes. We can skip to mark a node with zero generation to be alive.

I see a contradiction now, and this highlights that there are actually two problems we try to solve here. The contradiction is this: If we're fine with overriding endpoint states on the other nodes, and we only want to prevent the replacing node to be marked alive, then we don't need to gossip with fake generation, we can gossip with the standard generation and only have the replacing node not respond to those requests when it's in the shadow mode.

If we do want to prevent overriding the old state on other nodes, then generation 0 approach is not enough, as it may override the old state in some cases.

I don't know if we want to prevent overriding the state. It seems safer to not override it, but I can't think of any problem with overriding either.

> 
> > > > Not sending responses seems like a safer way to implement a shadow round.
> > > 
> > > 
> > > With this patch, we do not need to introduce another gossip mode. It runs the normal gossip mode. It is safer in the sense we do not need to worry about the regression introduced by this new long shadow round mode.
> > 
> > 
> > The `advertise_myself` mode is also a new mode which can cause regressions. I'm not sure why you think the long shadow round will be less safe. The long shadow round only suppresses messages. Other nodes will perceive it the same way as if the node was partitioned. Sending no messages seems safer than actually sending messages but with a special generation number, because it doesn't affect the state of other nodes.
> 
> No, this patch introduces no new mode. The gossip protocol is not changed. Look at the gossip handlers, nothing has changed there in this patch. No special handling for shadow round. This patch only changes the way the node choose the gossip generation.

Suppressing messages is not exactly changing the protocol, other nodes won't notice anything. Why do you dislike changing the handlers?

> 
> It is not sending any message in most of the cases because other nodes have bigger generation. In special cases when other nodes also think the node has a zero generation, this patch will still work even with this version, because echo message to this replacing node will fail. It is even more safe with v2 below, because other nodes will not try to mark a node with zero generation at all.
> 
> More importantly, this patch fixes an real bug in your proposal - not sending message is not enough.

Yes, but the same fix can be applied to my proposal as well, it doesn't invalidate it.

Reply:
I used advertise_myself_t because I want to use advertise_myself as the variable name.

Reply:
But I found a way to solve this and drop the _t suffix.

Reply:
You can still use it both as a type and the parameter name.  For example:
```
enum class yes_or_no { no, yes };

bool check(yes_or_no yes_or_no) {
    return yes_or_no == yes_or_no::yes;
}
```